### ISSUE TITLE
Use GPU when working with TPOT-NN

### ISSUE THREAD

**Issue by sword134** [EpistasisLab/tpot#1102]
How does one correctly use a cuda enabled GPU with the TPOT-NN configuration? 
Does the TPOT-NN configuration use a NN if you specify the code as follows or does one need to specify a template as well?

```
tpot = TPOTClassifier(generations=None, population_size=250, verbosity=2, max_time_mins=60, n_jobs=-1, config_dict="TPOT NN")
tpot.fit(X_train, y_train)
print(tpot.score(X_val, y_val))
```

**Comment by weixuanfu** [comment]
@JDRomano2 Could you please answer this question? Thank you.

**Comment by JDRomano2** [comment]
The NN models should automatically use GPU when it's available:
https://github.com/EpistasisLab/tpot/blob/219f8c5abe43996abb2c19d6a1767083304a23d3/tpot/builtins/nn.py#L260

However, we haven't yet written any explicit tests to ensure that this works as intended. Importantly, you need to have installed a CUDA-enabled version of PyTorch as well. I'll report back when I've established whether it works as intended when both (a.) a CUDA compatible GPU is available on the system and (b.) a CUDA version of PyTorch is installed.

**Comment by JDRomano2** [comment]
Confirmed that CUDA is used when available, under the following testing conditions:

- Windows 10
- PyTorch 1.6.0
- CUDA Toolkit 10.1
- Nvidia GeForce GTX 1080

![2020-08-14 12_00_04-Task Manager](https://user-images.githubusercontent.com/843449/90269780-2d1a2380-de27-11ea-991c-97ca8268b326.png)

My testing script:

```python
import tpot
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split

X, y = make_classification(n_samples=1000, n_features=20, n_informative=4, random_state=42)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

clf = tpot.TPOTClassifier(generations=50, population_size=10, verbosity=2, config_dict='TPOT NN', template='Selector-Transformer-PytorchMLPClassifier')

clf.fit(X_train, y_train)
print(clf.score(X_test, y_test))
```

@sword134 - would you mind verifying your installed PyTorch version, CUDA version, and [GPU compute capability](https://developer.nvidia.com/cuda-gpus#compute)?

**EDIT**: This works both for the template and non-template ways of running TPOT.

**Comment by sword134** [comment]
@JDRomano2 Very well I have CUDA and pytorch and cuda is enabled when running it through pytorch. Now I just want my other question answered.
"Does the TPOT-NN configuration use a NN if you specify the code as follows or does one need to specify a template as well?"

**Comment by JDRomano2** [comment]
**tl;dr:** No, you don't need to specify a template to use NN estimators, but you can if you want.

When you say `config_dict='TPOT NN'`, all that does is make the NN estimators available to TPOT. It doesn't necessarily guarantee that TPOT will include them in the final pipeline, but they _can_ be in the pipeline, along with any of the other non-NN estimators that are built in.

The template option is good for if you know that (a.) you definitely want an NN estimator and (b.) you already know what format pipeline you want and just need to optimize hyperparameters.

If you still want TPOT to learn the pipeline format, but also want to restrict it to **only** use NN estimators in that pipeline, you can specify a custom config dict, something like the following:

```python
clf = tpot.TPOTClassifier(
    generations=50,
    population_size=10,
    verbosity=2,
    config_dict={
    'tpot.builtins.PytorchLRClassifier': {
        'learning_rate': [1e-3, 1e-2, 1e-1, 0.5, 1.],
        'batch_size': [4, 8, 16, 32],
        'num_epochs': [5, 10, 15],
        'weight_decay': [0, 1e-4, 1e-3, 1e-2]
    },

    'tpot.builtins.PytorchMLPClassifier': {
        'learning_rate': [1e-3, 1e-2, 1e-1, 0.5, 1.],
        'batch_size': [4, 8, 16, 32],
        'num_epochs': [5, 10, 15],
        'weight_decay': [0, 1e-4, 1e-3, 1e-2]
    },
    }
)
```

**Comment by weixuanfu** [comment]
Thank you @JDRomano2 I close the issue. Please feel free to reopen it if there are other related questions.

**Comment by sword134** [comment]
Running the following code gives me an error when calling tpot.score

```
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, shuffle=False)
X_val, X_test, y_val, y_test  = train_test_split(X_test, y_test, test_size=0.50, random_state=1, shuffle=False) # 0.5 x 0.33 = 0.165

tpot = TPOTClassifier(generations=None, population_size=1000, verbosity=2, max_time_mins=60, n_jobs=-1, periodic_checkpoint_folder="D:\\ReinforcementLearning\\XGBoost\\Output\\Checkpoints", config_dict="TPOT NN")
tpot.fit(X_train, y_train)
tpot.export('D:\\ReinforcementLearning\\XGBoost\Output\\tpot_digits_pipeline.py')
print(tpot.score(X_val, y_val))
```

```
Traceback (most recent call last):
  File "d:\ReinforcementLearning\XGBoost\XGBoost.py", line 43, in <module>
    print(tpot.score(X_val, y_val))
  File "D:\Anaconda\envs\NBA\lib\site-packages\tpot\base.py", line 957, in score
    testing_target.astype(np.float64)
  File "D:\Anaconda\envs\NBA\lib\site-packages\pandas\core\generic.py", line 5537, in astype
    new_data = self._mgr.astype(dtype=dtype, copy=copy, errors=errors,)
  File "D:\Anaconda\envs\NBA\lib\site-packages\pandas\core\internals\managers.py", line 567, in astype
    return self.apply("astype", dtype=dtype, copy=copy, errors=errors)
  File "D:\Anaconda\envs\NBA\lib\site-packages\pandas\core\internals\managers.py", line 396, in apply
    applied = getattr(b, f)(**kwargs)
  File "D:\Anaconda\envs\NBA\lib\site-packages\pandas\core\internals\blocks.py", line 590, in astype
    values = astype_nansafe(vals1d, dtype, copy=True)
  File "D:\Anaconda\envs\NBA\lib\site-packages\pandas\core\dtypes\cast.py", line 989, in astype_nansafe
    return arr.astype(dtype, copy=True)
ValueError: could not convert string to float: 'Bull'
```

**Comment by JDRomano2** [comment]
This looks like a formatting issue with your dataset, not with TPOT. You may need to map your `y` values to integers.
